Microsoft.Identity.Web/Client/TokenCacheProviders/Session/MSALPerUserSessionTokenCacheProvider.cs

```diff
@@ -0,0 +1,233 @@
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.Identity.Client;
+using System.Diagnostics;
+using System.Security.Claims;
+using System.Threading;
+
+namespace Microsoft.Identity.Web.Client.TokenCacheProviders
+{
+    /// <summary>
+    /// This is a MSAL's TokenCache implementation for one user. It uses Http session as a backend store
+    /// </summary>
+    public class MSALPerUserSessionTokenCacheProvider : IMSALUserTokenCacheProvider
+    {
+        private static ReaderWriterLockSlim SessionLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+
+        /// <summary>
+        /// Once the user signes in, this will not be null and can be ontained via a call to ClaimsPrincipal.Current
+        /// </summary>
+        internal ClaimsPrincipal SignedInUser;
+
+        /// <summary>
+        /// The HTTP context being used by this app
+        /// </summary>
+        internal HttpContext HttpContext = null;
+
+        /// <summary>
+        /// The internal handle to the client's instance of the Cache
+        /// </summary>
+        private ITokenCache UserTokenCache;
+
+        /// <summary>Initializes a new instance of the <see cref="MSALPerUserSessionTokenCache"/> class.</summary>
+        public MSALPerUserSessionTokenCacheProvider()
+        {
+        }
+
+        /// <summary>Initializes the cache instance</summary>
+        /// <param name="tokenCache">The ITokenCache passed through the constructor</param>
+        /// <param name="httpcontext">The current HttpContext</param>
+        /// <param name="user">The signed in user's ClaimPrincipal, could be null.
+        /// If the calling app has it available, then it should pass it themselves.</param>
+        public void Initialize(ITokenCache tokenCache, HttpContext httpcontext, ClaimsPrincipal user)
+        {
+            this.HttpContext = httpcontext;
+
+            this.UserTokenCache = tokenCache;
+
+            this.UserTokenCache.SetBeforeAccess(this.UserTokenCacheBeforeAccessNotification);
+            this.UserTokenCache.SetAfterAccess(this.UserTokenCacheAfterAccessNotification);
+            this.UserTokenCache.SetBeforeWrite(this.UserTokenCacheBeforeWriteNotification);
+
+            if (user == null)
+            {
+                // No users signed in yet, so we return
+                return;
+            }
+
+            this.SignedInUser = user;
+            this.LoadUserTokenCacheFromSession();
```

Comment:
I don't think that's needed. It will be loaded just in time.

Reply:
Added this after i found out during my testing that the user's token cache was not hit unless they triggered a sign-on. This helps when the user didn't sign out and starts the application again

---
In reply to: [267389622](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/65#discussion_r267389622) [](ancestors = 267389622)

Reply:
I don't understand.

Reply:
So when the user signs in, exits the browser without signing off, and signs in again, this code loads the token cache for the user again. If this was not there, the token cache is not hit, till the user presses the sign-in button again. This ensures that the cache is definitely hit if the claims principal is present and active. This is more noticable in unot testing (daemon, console  scenarios). Not so much in web though, where most of the magic is done via cookies. Since the idea is that "Session" cache will only be mostly used on web, this can be dropped from Sessionbased user caches. 

---
In reply to: [268398315](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/65#discussion_r268398315) [](ancestors = 268398315)